### ISSUE TITLE
Time Filter Enhacement

### ISSUE THREAD

**Issue by federicokeen** [metabase/metabase#3199]
This is a feature request, not a bug. 

Kibana or Grafana have very good date filters. I base my suggestion in their options. Basically right now to be able to filter a dashboard by a date I had to add 3 time filters on the same field, to provide Relative, Month and Year or Quarter selection. Missing very much the possibility to define a specific range there as well. For example I needed to do the last 6 months, and I couldn't since its not a predefined relative option and I cannot specify a range. 

I'm attaching screenshots from Kibana and Grafana as examples:

**Kibana**: 
<img width="1440" alt="screen shot 2016-08-16 at 12 12 33 am" src="https://cloud.githubusercontent.com/assets/2558719/17682892/94de27ce-6346-11e6-952a-4fcbf9e95f45.png">
<img width="1422" alt="screen shot 2016-08-16 at 12 12 05 am" src="https://cloud.githubusercontent.com/assets/2558719/17682903/9fff6d98-6346-11e6-9cd3-59b4e5055e07.png">
<img width="1440" alt="screen shot 2016-08-16 at 12 12 13 am" src="https://cloud.githubusercontent.com/assets/2558719/17682904/a6c9b5ca-6346-11e6-9c89-ef7bdcd161de.png">
<img width="1432" alt="screen shot 2016-08-16 at 12 12 21 am" src="https://cloud.githubusercontent.com/assets/2558719/17682905/ad24263a-6346-11e6-92c1-f7633c39084e.png">

**Grafana**:
<img width="1440" alt="screen shot 2016-08-16 at 12 12 52 am" src="https://cloud.githubusercontent.com/assets/2558719/17682914/c2b17a2a-6346-11e6-8103-90d07280a147.png">
<img width="1438" alt="screen shot 2016-08-16 at 12 13 00 am" src="https://cloud.githubusercontent.com/assets/2558719/17682916/c62478c4-6346-11e6-9d3c-7e87a93a5781.png">

**Comment by moumny** [comment]
I believe this issue is mostly about the dashboards. In general the Timestamp filter widget in the query builder is not bad and should be available in the dashboards(That is having both relative and absolute options). In the dashboards the single date or date range filter cannot have a reasonable defaults like "last 7 days".

Then there is the question about adding more relative filter options. For example there is no way to filter on this or last quarter / 6 months. So either adding all the options or having a custom relative filter widget for things like "past X [hour, day, week, month, year]"

Last thing, and maybe this is out of scope here, but for all timeseries-like data it makes sense to control both the filter and the group by time granularity in the same place. For example we have hourly line chart in dashboard with relative 7 days filter, but if user selects 1 year then it creates a mess.(it should be possible to get the data in daily or weekly)

**Comment by tlrobinson** [comment]
@iwaffles was also requesting additional date filters to me the other day, specifically "next 3 days", but IMHO we should just allow arbitrary intervals since we can't predict exactly how people want to use it.

For dashboards, I think it's nice to have some simple widgets like we currently have, but I also think you should be able to access the full power of the query builder date widget if you want to. Either as a separate widget option, or maybe a small button at the bottom of the other date widgets that allow you to drop into the full date widget.

Regarding combining the filter and grouping granularity, that's an interesting idea. At one point we considered having a separate area of the query builder specifically for the time dimension. Perhaps if you group by a date column we could show an extra timeseries specific widget that allows you to control the range and granularity.

**Comment by moumny** [comment]
Heres a few benefits i see in having time filter and grouping together:

-Most of the time you can select reasonable defaults in granularity by looking at the time range
- You can prevent things like querying hourly data for past year (or at least by default move the granularity to something reasonable)
- No need to have hourly daily weekly version of each question (So less questions, we discussed this in the question organization issue)
-Suppose you have an hourly timeseries and default have time filter set  to 7 days. now if you move to 30 days or year then the graph is unreadable
-Also it makes sense to be able to configure default filter (For big databases would be nice not to query TBs of data every time ATM, we do that to add questions to dashboard without filter)
-I'm not sure how frequent is it query a table with a timestamp column without a time filter. usually this is "filter since i started populating this table ^^"
-For not very advanced users(we have some of those) it would make the query builder a little bit less like "writing an SQL query with a GUI" and a bit more like "ask a question". As they can just select a period, a metric and play around a little bit.

tl,dr less clicks, less questions, better dashboards, easier QB

Basically all this makes me think its worth giving timestamp a VIP treatment from the rest of the columns.

**Comment by kdoh** [comment]
Hey all.

Wanted to share some of our thinking on an updated time filter widget. The main goal here is to expand the expressivity of the filter widget while still making time filtering quick and as intuitive as possible. 

Here's a short list of the main features we want to add based on all the great feedback from all of you here and elsewhere:
- The ability to express arbitrary relative ranges. E.x. 180 days ago.
- The ability to express relative ranges into the future. (Imagine reservations, bookings, etc)
- The ability to add a specific times. 
- A larger range of time granularities for time filtering. (Seconds, Hours, etc).

The wireframes I'm sharing are mostly about showcasing the overall flow and functionality of the widget so it's important to remember this isn't necessarily indicative of the final visual style. :) 

Here we see an existing filter set to the current week. (That would be "This week" in the current time filter parlance)

![filter_01](https://cloud.githubusercontent.com/assets/5248953/19783921/b0c5aaf8-9c48-11e6-900b-3ffa8d538856.png)

If I want to change the type of filter I'm applying, I can click and change from Current to the other types of time filter options. In this case I want to do a little bit of work in the future so I'll choose "Next".

![filter_02](https://cloud.githubusercontent.com/assets/5248953/19783937/bde74fe8-9c48-11e6-8485-bbe169fbbaea.png)

I can enter any numeric value I want here. For this particular case I want to look at the next 24 hours so I'll enter 24. 

![filter_03](https://cloud.githubusercontent.com/assets/5248953/19783946/c3ea38a6-9c48-11e6-915c-a79ac61ade6a.png)

Since the granularity is set to days by default, but I want hours, I'll now switch the granularity to hours. Here you can see the new set of granularities we're enabling. 

![filter_04](https://cloud.githubusercontent.com/assets/5248953/19783963/d25c8010-9c48-11e6-8421-1499f13cbaf9.png)

Now I'm going to switch gears and apply a time range. In this case I want to always see everything that happened after a specific date. I'll choose after and now I'm presented with a specific date.

![filter_05](https://cloud.githubusercontent.com/assets/5248953/19783975/d8b79e0e-9c48-11e6-8f1e-a30141233299.png)

If I click on the date I'm presented with a calendar picker. While picking the date I want I can either type a date or use the calendar to update the value.

![filter_06](https://cloud.githubusercontent.com/assets/5248953/19783986/e25b4f00-9c48-11e6-9c03-ed631008194d.png)

But in addition to seeing everything that happened after 10/21/2016, I specifically want to look at everything that happened after a specific time on that day. You may have noticed the "Pick a time" trigger in the original "After" screenshot. If I click on this I'm presented with an additional input where I can specify the time I want.

![filter_07](https://cloud.githubusercontent.com/assets/5248953/19784017/04bac77e-9c49-11e6-9c8f-b75413c5ce30.png)

If i want to select a time range in between two specific days I can pick "Between" and I'm presented with two of the date inputs. All of the behavior is the same here and I can also specify time values here as well.

![filter_08](https://cloud.githubusercontent.com/assets/5248953/19784031/0d28e5c6-9c49-11e6-831e-352bf1987410.png)

Would love to hear thoughts on this new approach!

**Comment by mazameli** [comment]
Nice work, Kyle! Some thoughts:

![b0c5aaf8-9c48-11e6-900b-3ffa8d538856](https://cloud.githubusercontent.com/assets/2223916/19785689/3fba83a4-9c4f-11e6-9506-f374058f9634.png)
`Current` and `Week` look to me like two unrelated possible selections rather than a single filter. I wonder if putting them side-by-side so that they read like a sentence would help; maybe also lowercasing "week." Not sure how far you are on visual design, but it also feels like a lot of empty white space to me. I know Tom gets wigged out by menus resizing, but if we can implement some CSS-transition-esque height/width transitions, I think it looks sleeker to fit the menus to the content.

![bde74fe8-9c48-11e6-8485-bbe169fbbaea](https://cloud.githubusercontent.com/assets/2223916/19785802/c2dd09fa-9c4f-11e6-9dc7-37c33d067d8f.png)
I dig this. Some nitpicky things: the `Current…` example text includes "Day," but now that `Today` is its own selection, it feels weird that there could also be `Current day`. Also, "Month" should be lowercase, and `etc` should have a period.

![c3ea38a6-9c48-11e6-915c-a79ac61ade6a](https://cloud.githubusercontent.com/assets/2223916/19785870/21397b6e-9c50-11e6-9d55-790f5d5662f9.png)
Similar comments as on the first image: the arrangement of these elements makes it a little hard for me to understand how they're related. Again, I wonder if a horizontal lockup would work better. I'm also not sure what making the `24` so large compared to the text is doing for us; is it intended or appropriate to make it the primary read? If it were the same size as the dropdown labels, but still purple, doesn't that make it stand out enough? Seems to me that each element in that "phrase" is equally important/consequential.

![d25c8010-9c48-11e6-8421-1499f13cbaf9](https://cloud.githubusercontent.com/assets/2223916/19786008/b0e8b360-9c50-11e6-8b63-6a7953b37193.png)
Feels like we shouldn't need to crop the rest of the content of the menu.

![d8b79e0e-9c48-11e6-8f1e-a30141233299](https://cloud.githubusercontent.com/assets/2223916/19786047/dfa98184-9c50-11e6-8d8a-b53b60c229f1.png)
The `After` label feels like it could be closer to the date to me. Details: the little dropdown arrow is out of place, and there's an errant arrow inside the purple button.

![e25b4f00-9c48-11e6-9c03-ed631008194d](https://cloud.githubusercontent.com/assets/2223916/19786133/442b9714-9c51-11e6-8c15-04b85863d6b0.png)
Having the date input field displayed simultaneously with the calendar picker feels sort of uncanny to me. Almost like I'd rather have it be a text field that I can type into, but with a calendar icon button next to it that I could click to invoke _and replace_ the input field. Part of what is feeling weird is that the input field says one thing while the calendar label says another, but also that it's giving me two simultaneous ways to change the value. Also, the stroke below the calendar/above the button seems needless.

![04bac77e-9c49-11e6-9c8f-b75413c5ce30](https://cloud.githubusercontent.com/assets/2223916/19786256/dbdc75ec-9c51-11e6-8103-cdb50b0299af.png)
I think you could get away with killing the border around `AM` and `PM`.

![0d28e5c6-9c49-11e6-831e-352bf1987410](https://cloud.githubusercontent.com/assets/2223916/19786315/3ac81552-9c52-11e6-85d5-c841e6c966b1.png)
Again, I think the `Between` label feels far away from the thing it's related to (the date), but also it feels weird that it's above instead of to the left of the first date, since the dates and the `and` all read in a straight line. Something about the placement/sizing of the `add filter` button isn't quite jiving for me in this state, either.

Lastly, one thing that might be missing (intentionally?) is the `is empty` and `is not empty` options that we currently have.

Hope this is helpful. I think this will be a great improvement on what we currently have!

**Comment by salsakran** [comment]
this was done in #3750
